### PR TITLE
Backend user validation

### PR DESCRIPTION
**Pull request by Metemi**
Create validation for user in DTO

### CODE REVIEW

**Review by markkovari — COMMENTED** [review]
This is looking fantastic, but I want you to go even further and add one happy and unhappy case at least as tests. 🥺 Please

**Comment by Metemi** [comment]
https://images.app.goo.gl/uV5KnKGHpNKJ2Hc28

**Comment by Metemi** [comment]
Well I got 403, I will try fixed it

File: Backend-API/src/test/java/com/avangers/backendapi/controllers/UserControllerTest.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    private UserServiceImpl userServiceImpl;
+
+    @Autowired
+    private MockMvc mockMvc;
+
+    @Autowired
+    private PasswordEncoder passwordEncoder;
+
+    //      objectMapper is used to convert objects to JSON
+    private ObjectMapper objectMapper;
```
Comment (markkovari) [inline Backend-API/src/test/java/com/avangers/backendapi/controllers/UserControllerTest.java:36]:
```suggestion
    //      objectMapper is used to convert objects to JSON
    @Autowired
    private ObjectMapper objectMapper;
```

If I am not mistaken you might even can autowire it. Not 100%.

File: Backend-API/src/test/java/com/avangers/backendapi/controllers/UserControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    void shouldRegisterUserWithCorrectNameAndPassword() throws Exception {
+        RegisterUserDTO validUser = new RegisterUserDTO("user@example.com", "Abc123456");
+        given(userServiceImpl.addUser(validUser)).willReturn(ResponseEntity.ok("Registration was successful"));
+
+        // should return 200 OK
+        mockMvc.perform(post("/register")
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .content(objectMapper.writeValueAsString(validUser)))
+                .andExpect(status().isOk());
+    }
```
Comment (markkovari) [inline Backend-API/src/test/java/com/avangers/backendapi/controllers/UserControllerTest.java:54]: https://www.digitalocean.com/community/tutorials/junit-display-name-displayname you can use the Displayname annotation too.

```suggestion
    @Displayname("should return 200 OK if request is valid")
    @Test
    void shouldRegisterUserWithCorrectNameAndPassword() throws Exception {
        RegisterUserDTO validUser = new RegisterUserDTO("user@example.com", "Abc123456");
        given(userServiceImpl.addUser(validUser)).willReturn(ResponseEntity.ok("Registration was successful"));

        mockMvc.perform(post("/register")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(validUser)))
                .andExpect(status().isOk());
    }
```

File: Backend-API/src/test/java/com/avangers/backendapi/controllers/UserControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+    @Test
+    void shouldNotRegisterWithBadPassword() throws Exception {
+        RegisterUserDTO userWithBadPassword = new RegisterUserDTO("user@example.com", "badpassword");
+        given(userServiceImpl.addUser(userWithBadPassword)).willReturn(
+                ResponseEntity.badRequest().body(
+                        "Password should contain at least one uppercase and one lowercase letter"));
+
+//        should return 400 BAD REQUEST
```
Comment (markkovari) [inline Backend-API/src/test/java/com/avangers/backendapi/controllers/UserControllerTest.java:63]: Same here.

File: Backend-API/src/test/java/com/avangers/backendapi/controllers/UserControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    @Autowired
+    private MockMvc mockMvc;
+
+    @Autowired
+    private PasswordEncoder passwordEncoder;
+
+    //      objectMapper is used to convert objects to JSON
+    private ObjectMapper objectMapper;
+
+    //      setUp method is used to create the objectMapper before each test
```
Comment (markkovari) [inline Backend-API/src/test/java/com/avangers/backendapi/controllers/UserControllerTest.java:38]: Maybe it is just me, but I feel like, your comments are "just describing" the 4 lines below. It is almost exactly as the code. Just a heads-up; if you want to add this comment you commit to maintain it too.

**Review by markkovari — CHANGES_REQUESTED** [review]
Just a few nits remanining.

**Comment by markkovari** [comment]
@Metemi just a little bit of help.

Since we opted in to use spring security, we have 403's everywhere. We need to enable requests to work without authentication. 





And here is an example too:
https://github.com/gfa-cc-after/Pet-Clinic-GilmoreDevs/blob/b5b403661ae1b848b26e0f4164db6c9b60704b93/backend/src/main/java/com/greenfoxacademy/backend/config/SecurityConfig.java#L26-L35

**Comment by Metemi** [comment]
Thank you Mark I will look at it

**Review by markkovari — APPROVED** [review]

**Review by MikeMoudry — APPROVED** [review]
image

**Pull request merged**